lua/ttt2/libraries/classbuilder.lua

```diff
@@ -0,0 +1,82 @@
+---
+-- A builder to handle the creation of classes from files
+-- @author Mineotopia
+-- @module classbuilder
+
+if SERVER then
+	AddCSLuaFile()
+end
+
+local tableCopy = table.Copy
+local tableDeepInherit = table.DeepInherit
+local stringSplit = string.Split
+local stringLower = string.lower
+local stringSub = string.sub
+local isfunction = isfunction
+
+classbuilder = classbuilder or {}
+
+---
+-- Builds a class from a file in a given scope. Uses the @{fileloader} to load files from a folder
+-- that are loaded as classes. Supports inheriting from base class.
+-- @note Has to be run on both server and client for client and shared files.
+-- @param string path The absolute path to search in, has to end with `/`
+-- @param[default=SHARED_FILE] number realm The realm where the file should be included
+-- @param string scope The scope where the new class will be registered, for example `ITEM`
+-- @param[opt] function OnInitialization This callback function is called on initialization of the class
+-- @param[default=false] boolean shouldInherit Set this to true if this class should inherit from its base
+-- @oaram[opt] function SpecialCheck A function that makes a special check, inheritance is blocked if false is returned
+-- @return table Returns a table of all the created classes
+-- @realm shared
+function classbuilder.BuildFromFolder(path, realm, scope, OnInitialization, shouldInherit, SpecialCheck)
+	-- In case this function is run on the server but the class should only exist
+	-- ono the client, this function should work only as a proxy.
+	if SERVER and realm == CLIENT_FILE then
+		fileloader.LoadFolder(path, false, realm)
+
+		return
+	end
+
+	-- In the first step make sure to cache the scope variable to keep
+	-- compatibility with addons that may use the same scope.
+	local cachedScope = _G[scope]
+
+	-- Now a global scope has to be created.
+	_G[scope] = {}
+
+	-- This variable will be used to store the loaded classes from the folder.
+	local classTable = {}
+
+	fileloader.LoadFolder(path, false, realm, function(filePath)
+		-- get the filename from the file path
+		local pathArray = stringSplit(filePath, "/")
+		local name = stringLower(stringSub(pathArray[#pathArray], 0, -5))
```

Comment:
Wouldn't run a `.lua` check and `.` split better?

Reply:
I thought about this as well. But you/ttt used this method in the previuos functions. using a dot to split might cause problems with `file.bla.lua`

Reply:
Agreed, BUT anyways, that's the file's name in that case: file.bla. So you just have to merge every split except the last one.

Reply:
Yes. But it would be worse performancewise

Reply:
Those files are loaded on startup, so that doesn't really matter. Additionally, files with an ending <> 3 chars would break this while the "engine" tries to load these files. That's just a stability thing tbh.